addons/mail/static/src/discuss/message_pin/common/message_model_patch.js

```diff
@@ -0,0 +1,76 @@
+import { patch } from "@web/core/utils/patch";
+import { Message } from "@mail/core/common/message_model";
+import { Record } from "@mail/core/common/record";
+import { _t } from "@web/core/l10n/translation";
+import { MessageConfirmDialog } from "@mail/core/common/message_confirm_dialog";
+import { Message as MessageComponent } from "@mail/core/common/message";
+
+patch(Message.prototype, {
+    setup() {
+        super.setup();
+        /** @type {luxon.DateTime} */
+        this.pinnedAt = Record.attr(undefined, { type: "datetime" });
+    },
+    update(data) {
+        super.update(data);
+        const { pinned_at: pinnedAt } = data;
+        if (this.thread?.model === "discuss.channel" && (pinnedAt !== undefined || this.isEmpty)) {
+            if (pinnedAt && !this.isEmpty) {
+                this.thread.pinnedMessages.add(this);
+                this.pinnedAt = pinnedAt;
+            } else {
+                this.pinnedAt = undefined;
+                this.thread.pinnedMessages.delete(this);
+            }
+        }
+    },
```

Comment:
Should turn `pinnedAt`/`pinnedMessages` into computed fields.

`update()` is not necessarily invoked on record update, this is just a syntactic sugar for `Object.assign(record, data)`. Having code relying on it being called is very weak, so that's not ok